### PR TITLE
Implement conditional widgets

### PR DESCRIPTION
**Pull request by ianroberts**
This is a first pass at implementing the "conditional" part of #164 - I've added an `if` option to the project config language, whose value is an expression that is evaluated against the current document data and the current state of the annotation data for that document, allowing widgets to decide whether they should be displayed based on the values set in other widgets.  For example:

```json
[
  {
    "name": "Document",
    "text": "{{{text}}}",
    "type": "html"
  },
  {
    "name": "sentiment",
    "type": "radio",
    "title": "Sentiment",
    "options": [
      "negative",
      "neutral",
      "positive"
    ]
  },
  {
    "type": "radio",
    "title": "Confidence",
    "name": "confidence",
    "options": [
      {"value":"1", "label": "1 (Not at all confident)"},
      "2","3","4",
      {"value": "5", "label": "5 (dead certain)"}
    ]
  },
  {
    "if": "annotation.confidence < 4",
    "name": "reason",
    "type": "text",
    "regex": "...",
    "title": "What makes you unsure?"
  }
]
```

In this example the "reason" field only appears if you select confidence level 1, 2 or 3, and the way I've done it the validation only checks visible components, so in this case the `regex` check on `reason` (requiring at least three characters) doesn't block submission when `confidence` is set to 4 or 5.

~~I've used the [jexl](https://www.npmjs.com/package/jexl) expression library but the last release of that looks like it was several years back, and there's at least one open issue on there that says it's vulnerable to prototype pollution attacks, so it may be worth looking for a different expression parser,~~ Edit: I've switched to [jse-eval](https://www.npmjs.com/package/jse-eval), which explicitly denies any access to `__proto__` and related properties.

I thought I'd submit this now so you can see what you think of the idea in principle.  Tasks still to do include:

- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by ianroberts** [comment]
@YLi999 would this be sufficient to cover your use case?

File: frontend/src/components/AnnotationRenderer.vue [hunk 1]
```diff
@@ -124,8 +126,23 @@ export default {
...
       return null
+    },
+    shownElements() {
+      if (!this.config) {
+        return [];
+      }
+      return this.config.filter((elemConfig) => {
+        if (this.condition[elemConfig.name]) {
+          try {
+            return !!this.condition[elemConfig.name].evalSync({document: this.document, annotation: this.annotationOutput});
```
Comment (ianroberts) [inline frontend/src/components/AnnotationRenderer.vue:137]: I've used `annotation` for the current annotation state in the expression bindings, but possibly a name like `values` might make more sense.

File: frontend/src/components/AnnotationRenderer.vue [hunk 2]
```diff
@@ -154,8 +171,17 @@ export default {
...
         for (let elemConfig of config) {
           this.validation[elemConfig.name] = null
+          if (elemConfig["if"]) {
+            try {
+              this.condition[elemConfig.name] = jexl.compile(elemConfig["if"])
+            } catch (_) {
+              // error compiling expression -> treat the same as if the element
+              // did not have an "if" at all, and always show it.
+            }
+          }
```
Comment (ianroberts) [inline frontend/src/components/AnnotationRenderer.vue:184]: The point of doing this in here is so we just _compile_ the expressions once when the page loads, and can later _evaluate_ them repeatedly against different contexts.

**Comment by ianroberts** [comment]
https://user-images.githubusercontent.com/428107/234087270-90a44185-b074-4ffd-9967-bf1016f896be.mov

File: frontend/src/components/AnnotationRenderer.vue [hunk 3]
```diff
@@ -124,8 +150,30 @@ export default {
         return  getValueFromKeyPath(this.document, this.doc_preannotation_field)
       }
       return null
+    },
+    shownElements() {
+      if (!this.config) {
+        return [];
+      }
+      return this.config.filter((elemConfig) => {
```
[the remark this answers is not in the document]
Comment (twinkarma) [reply frontend/src/components/AnnotationRenderer.vue:158]: Ah sorry it was an array built in, silly me

[the remark this answers is not in the document]
Comment (ianroberts) [reply frontend/src/components/AnnotationRenderer.vue:158]: I realise that we probably also need to filter the final `annotationOutput` to only include the values corresponding to the visible widgets, so you don't get the situation where someone sets confidence to 3, adds a reason, then changes their mind and modifies confidence to 4, the reason field disappears but the result json still has a `reason` property.

**Comment by twinkarma** [comment]
It's looking good so far. Focusing on getting vite in now and will have a proper look at this after.

**Comment by twinkarma** [comment]
When pressing `Clear` or `Submit` after having filled the form in, the form expands to show all options for some reason. Pressing `Clear` again resets the form to its original state.

https://user-images.githubusercontent.com/11231792/235942412-cf39e654-6110-48ed-aa26-17122d241f1b.mp4

File: frontend/src/components/AnnotationRenderer.vue [hunk 4]
```diff
@@ -46,6 +46,62 @@ import {DocumentType} from '@/enum/DocumentTypes';
...
+  // the most serious hole.
+  Object: Object.freeze(Object.assign(Object.create(null), {
+    entries: Object.entries,
+    fromEntries: Object.fromEntries,
+    keys: Object.keys,
+    values: Object.values,
+    getOwnPropertyNames: Object.getOwnPropertyNames,
+    hasOwn: Object.hasOwn,
+  }))
+}))
```
Comment (ianroberts) [inline frontend/src/components/AnnotationRenderer.vue:104]: This is the bit that has been bugging me slightly, and which I want to resolve one way or another before this PR is ready to merge.  Any time when we start accepting and evaluating arbitrary JS expressions from the (manager) user we need to think about security, and it's tricky to find the right balance between locking down things that are dangerous but still making the feature useful.

The expression parser is configured to reject assignment expressions but I realised that allowing `Object.assign` would bypass that.  And it blocks access to `prototype`, `__proto__` and `constructor` properties but `Object.getPrototypeOf` would bypass that...

The solution I've adopted here is to allow-list the static `Object.x` methods that are visible to `if` expressions, is that sufficient or do we need to do something similar with `Array` and `String`?  Would it be safer to block function expressions altogether, though they arere useful for things like `myArray.every((v) => v.something())`...

I don't give the expressions access to things like the global `window` object, but given `annotation` and `document` are not just plain objects but are enhanced with Vue's getters and setters for scope watching, does that provide a backdoor way to get at the DOM element that the component is attached to, and thence to things like `window.alert`?

Basically how paranoid are we/should we be?

Comment (ianroberts) [reply frontend/src/components/AnnotationRenderer.vue:104]: > given `annotation` and `document` are not just plain objects but are enhanced with Vue's getters and setters for scope watching, does that provide a backdoor way to get at the DOM element that the component is attached to, and thence to things like `window.alert`?

Why yes, yes it does...

```javascript
annotation.__ob__.dep.subs[0].vm.$el.ownerDocument.defaultView.alert('haha')
```

**Comment by ianroberts** [comment]
Following discussion with Twin I've switched to a new approach where we block _all_ function calls by default and introduce specific syntax to cover the cases where functions were previously necessary (principally any/all quantifiers over arrays - "show component B if _any_ of the values selected for checkbox A matches predicate P" - and testing of regular expressions against strings).  I believe this strikes a better balance between expressivity vs security, and I think this PR is ready for proper review now.

**Review by twinkarma — APPROVED** [review]
LGTM! Amazing work on this @ianroberts.

**Pull request merged**